### PR TITLE
Add methods for testing if a directory is writable

### PR DESCRIPTION
**Pull request by robie2011**

### CODE REVIEW

File: FileSystem.Test/FileSystemTest.cs [hunk 1]
```diff
@@ -14,6 +17,33 @@ public sealed class FileSystemTest
         private static readonly string FileInFolder = Path.Combine(Folder, "file.txt");
         private static readonly string FileInSubFolder = Path.Combine(SubFolder, "file.txt");
 
+        [Fact]
+        public void CheckDirectoryIsWritable()
+        {
+            if (Environment.OSVersion.Platform == PlatformID.Unix ||
```
Comment (janhohenheim) [inline FileSystem.Test/FileSystemTest.cs]: Use a [reusable custom fact attribute](https://stackoverflow.com/a/4421941/5903309) for this

File: FileSystem.Test/FileSystemTest.cs [hunk 2]
```diff
@@ -14,6 +17,33 @@ public sealed class FileSystemTest
         private static readonly string FileInFolder = Path.Combine(Folder, "file.txt");
         private static readonly string FileInSubFolder = Path.Combine(SubFolder, "file.txt");
 
+        [Fact]
+        public void CheckDirectoryIsWritable()
```
Comment (janhohenheim) [inline FileSystem.Test/FileSystemTest.cs]: Mention the fact that this only runs on Windows

File: FileSystem.Test/FileSystemTest.cs [hunk 3]
```diff
@@ -14,6 +17,33 @@ public sealed class FileSystemTest
...
+            var denyAllRule = new FileSystemAccessRule(windowsIdentity.Name, FileSystemRights.FullControl, AccessControlType.Deny);
+            var path = Path.Combine(testEnvironment.RootDirectory, Folder);
+            var directoryInfo = new DirectoryInfo(path);
+            Assert.True(filesystem.DirectoryIsWritable(path));
+
+            var security = directoryInfo.GetAccessControl();
+            security.AddAccessRule(denyAllRule);
+            directoryInfo.SetAccessControl(security);
+            var isWritable = filesystem.DirectoryIsWritable(path);
+            Assert.False(isWritable);
```
Comment (janhohenheim) [inline FileSystem.Test/FileSystemTest.cs:38]: You are testing two things here. Split this test.

File: FileSystem/FileSystem.cs [hunk 4]
```diff
@@ -28,6 +29,32 @@ public void Delete(string path)
             }
         }
 
+        public bool DirectoryIsWritable(string path)
```
Comment (janhohenheim) [inline FileSystem/FileSystem.cs]: Follow the conventions of other predicates and start the method with "Is": `IsDirectoryWritable`

File: FileSystem/FileSystem.cs [hunk 5]
```diff
@@ -28,6 +29,32 @@ public void Delete(string path)
...
+                {
+                    file.WriteLine("test");
+                }
+
+                Delete(dummyFilePath);
+                return true;
+            }
+            catch (Exception)
+            {
+                return false;
```
Comment (janhohenheim) [inline FileSystem/FileSystem.cs:54]: This will return false even if only the deletion fails.

File: FileSystem/IFileSystem.cs [hunk 6]
```diff
@@ -54,5 +54,12 @@ public interface IFileSystem
         /// <param name="searchPattern">The search string to match against the names of files in path. This parameter can contain a combination of valid literal path and wildcard (* and ?) characters, but it doesn't support regular expressions.</param>
         /// <returns>An IEnumerable of the full names (including paths) for the files in the specified directory.</returns>
         IEnumerable<string> GetFiles(string path, string searchPattern);
+
+        /// <summary>
+        /// Returns boolean that represents directory is writeable.
```
Comment (janhohenheim) [inline FileSystem/IFileSystem.cs]: This piece of documentation simply restates the method name and as such can be omitted.

File: FileSystem/IFileSystem.cs [hunk 7]
```diff
@@ -54,5 +54,12 @@ public interface IFileSystem
...
+        /// </summary>
+        /// <param name="path">The relative or absolute path to the directory to search. This string is not case-sensitive.</param>
+        /// <returns>A boolean that represents directory is writeable.</returns>
```
Comment (janhohenheim) [inline FileSystem/IFileSystem.cs]: Remove this line as well

File: FileSystem/IFileSystem.cs [hunk 8]
```diff
@@ -54,5 +54,12 @@ public interface IFileSystem
         /// <param name="searchPattern">The search string to match against the names of files in path. This parameter can contain a combination of valid literal path and wildcard (* and ?) characters, but it doesn't support regular expressions.</param>
         /// <returns>An IEnumerable of the full names (including paths) for the files in the specified directory.</returns>
         IEnumerable<string> GetFiles(string path, string searchPattern);
+
+        /// <summary>
+        /// Returns boolean that represents directory is writeable.
+        /// </summary>
+        /// <param name="path">The relative or absolute path to the directory to search. This string is not case-sensitive.</param>
```
Comment (janhohenheim) [inline FileSystem/IFileSystem.cs]: "not case-sensitive" -> "case-insensitive"

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: FileSystem.Test/FileSystemTest.cs (hunk 3 above)
Comment (robie2011) [reply FileSystem.Test/FileSystemTest.cs:38]: This is one test. The first part - including first assert - is preparation for actual testing.

File: FileSystem/IFileSystem.cs (hunk 6 above)
Comment (robie2011) [reply FileSystem/IFileSystem.cs]: Ok 9a84d40937ee86dfc454b13b69ac580354a23839

File: FileSystem/IFileSystem.cs (hunk 8 above)
Comment (robie2011) [reply FileSystem/IFileSystem.cs]: I have copied that from GetFiles method. Are you sure? What about unix/linux systems?

File: FileSystem/FileSystem.cs (hunk 5 above)
Comment (robie2011) [reply FileSystem/FileSystem.cs:54]: Yes, that was intentionally. 
I think in our context we want to write AND delete files in directory.
Isn't that necessary?

File: FileSystem/FileSystem.cs (hunk 4 above)
Comment (robie2011) [reply FileSystem/FileSystem.cs]: https://github.com/messerli-informatik-ag/file-system/pull/22/commits/cd7be418ffe2f73ac6cfdd9aa1b751314b673f9c

File: FileSystem.Test/FileSystemTest.cs (hunk 2 above)
Comment (robie2011) [reply FileSystem.Test/FileSystemTest.cs]: https://github.com/messerli-informatik-ag/file-system/pull/22/commits/cd7be418ffe2f73ac6cfdd9aa1b751314b673f9c

File: FileSystem.Test/FileSystemTest.cs (hunk 1 above)
Comment (robie2011) [reply FileSystem.Test/FileSystemTest.cs]: Good idea. I see there is already an Attribute called `SkippableFact` https://stackoverflow.com/a/35871507/2248405.
fb2c658

File: FileSystem/IFileSystem.cs (hunk 7 above)
Comment (robie2011) [reply FileSystem/IFileSystem.cs]: 9a84d40

File: FileSystem.Test/FileSystemTest.cs (hunk 1 above)
Comment (robie2011) [reply FileSystem.Test/FileSystemTest.cs]: test

Comment (janhohenheim) [reply FileSystem.Test/FileSystemTest.cs]: Good catch, feel free to use that NuGet package

File: FileSystem.Test/FileSystemTest.cs (hunk 3 above)
Comment (janhohenheim) [reply FileSystem.Test/FileSystemTest.cs:38]: Then rename the test to reflect expected failure

File: FileSystem/IFileSystem.cs (hunk 8 above)
Comment (robie2011) [reply FileSystem/IFileSystem.cs]: 5ba99b5

File: FileSystem.Test/FileSystemTest.cs (hunk 3 above)
Comment (robie2011) [reply FileSystem.Test/FileSystemTest.cs:38]: 0757299

**Review by janhohenheim — APPROVED** [review]

**Pull request merged**